src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs

```diff
@@ -15,6 +15,8 @@ namespace NuGet.Common
     public static class ConcurrencyUtilities
     {
         private const int NumberOfRetries = 3000;
+        // To maintain SHA-1 backwards compatibility with respect to the length of the hex-encoded hash, the hash will be truncated to a length of 20 bytes.
```

Comment:
Why do we need to maintain the 20 byte length? THe actual hashes are going to be different anyway, so I don't understand the benefit.

Reply:
As the SHA2 is 256 bits and SHA1 is 160 bits, the hash string will be 24 characters longer. The hash string will be part of the lock file path. So it's safer to truncate it as not to exceed the max length of path (260 characters)

Reply:
The 260 character path limit is Windows only, so we don't need to investigate Linux or Mac. This utility uses a base path of `%temp%\NuGetScratch\lock\`.

On my machine, running Windows 10this is `c:\users\<user_id>\AppData\Local\Temp\NuGetScratch\lock\`, which is 48 characters plus the length of `<user_id>`.

I believe on earlier versions of Windows, `\users\` is replaced with `\Documents and Settings\`, which makes it 65 characters plus the length of `<user_id>`.

Therefore, with a 40 character hash filename, `<user_id>` can be up to 172 or 155 characters, depending on the version of Windows, before the max path is reached.

If we used a 64 character hash filename, `<user_id>` can be up to 148 or 131 characters, depending on the version of Windows, before the max path is reached.

Admittedly, this is just for machines using the default location for the user profile path. If a machine sets the user profile to something that is longer than `c:\users\` or `c:\Documents and Settings\`, then the `<user_id>` segment needs to be shorter. But I think that allowing for a user id of well over 100 characters is sufficient.

Additionally, if we're concerned about the max path length, we should base64 encode the hash, not hex encode it. Hex encoding doubles the number of bytes, so 100% larger, whereas base64 is about 30% larger than the original number of bytes. Therefore base64 encoding a 32 byte hash will have about 2 characters more than hex encoding a 20 byte hash.

Reply:
Thanks for the investigation and detailed explanation!
For the base64 encoding, I think there're two problems:
1.The 64 chars are: a-z, A-Z, 0-9, + and /, / is illegal char for files and directories both on Windows and Linux.
2.Windows does not distinguish between upper-case and lower-case characters.

For the plugin cache, it uses two hash strings in the path https://github.com/NuGet/NuGet.Client/blob/4c9d2939dabec06a3b5f14e7cd8159bc59cd4179/src/NuGet.Core/NuGet.Protocol/Plugins/PluginCacheEntry.cs#L28-L29
So it looks like it will be more likely to exceed 260 charactors.